crates/bevy_asset/src/loader.rs

```diff
@@ -148,48 +149,84 @@ pub struct AssetResult<T: Component> {
 /// A channel to send and receive [AssetResult]s
 #[derive(Debug)]
 pub struct AssetLifecycleChannel<T: Component> {
-    pub sender: Sender<AssetLifecycleEvent<T>>,
-    pub receiver: Receiver<AssetLifecycleEvent<T>>,
+    pub to_system: Sender<AssetLifecycleEvent<T>>,
+    pub from_asset_server: Receiver<AssetLifecycleEvent<T>>,
 }
 
 pub enum AssetLifecycleEvent<T: Component> {
     Create(AssetResult<T>),
+    CreateFrom {
+        from: HandleId,
+        to: HandleId,
+        to_type_uuid: Uuid,
+        transform: Box<dyn (FnOnce(&T) -> Option<Box<dyn AssetDynamic>>) + Send>,
+    },
     Free(HandleId),
 }
 
 pub trait AssetLifecycle: Downcast + Send + Sync + 'static {
     fn create_asset(&self, id: HandleId, asset: Box<dyn AssetDynamic>, version: usize);
+    fn create_asset_from(
+        &self,
+        from: HandleId,
+        to: HandleId,
+        to_uuid: Uuid,
+        transform: Box<dyn (FnOnce(&dyn AssetDynamic) -> Option<Box<dyn AssetDynamic>>) + Send>,
+    );
     fn free_asset(&self, id: HandleId);
 }
 impl_downcast!(AssetLifecycle);
 
 impl<T: AssetDynamic> AssetLifecycle for AssetLifecycleChannel<T> {
     fn create_asset(&self, id: HandleId, asset: Box<dyn AssetDynamic>, version: usize) {
         if let Ok(asset) = asset.downcast::<T>() {
-            self.sender
+            self.to_system
                 .send(AssetLifecycleEvent::Create(AssetResult {
                     id,
                     asset: *asset,
                     version,
                 }))
                 .unwrap()
         } else {
-            panic!(
+            warn!(
                 "Failed to downcast asset to {}.",
                 std::any::type_name::<T>()
             );
         }
     }
 
+    /// Convert an asset from type `T` to type with `T_TO::TYPE_UUID == to_uuid`.
+    /// It is the responsibility of the caller to ensure that input/output of `transform`
+    /// matches those types
+    fn create_asset_from(
+        &self,
+        from: HandleId,
+        to: HandleId,
+        to_type_uuid: Uuid,
+        transform: Box<dyn (FnOnce(&dyn AssetDynamic) -> Option<Box<dyn AssetDynamic>>) + Send>,
+    ) {
+        self.to_system
+            .send(AssetLifecycleEvent::CreateFrom {
+                from,
+                to,
+                to_type_uuid,
+                transform: Box::new(|asset: &T| transform(asset)),
+            })
+            .unwrap();
+    }
+
     fn free_asset(&self, id: HandleId) {
-        self.sender.send(AssetLifecycleEvent::Free(id)).unwrap();
+        self.to_system.send(AssetLifecycleEvent::Free(id)).unwrap();
     }
 }
 
 impl<T: Component> Default for AssetLifecycleChannel<T> {
     fn default() -> Self {
-        let (sender, receiver) = crossbeam_channel::unbounded();
-        AssetLifecycleChannel { sender, receiver }
+        let (to_system, from_asset_server) = crossbeam_channel::unbounded();
+        AssetLifecycleChannel {
+            to_system,
```

Comment:
When reading through this I stumbled on the name `to_system` several times because it seems like it must be connected to the IntoSystem but that doesn't make sense in context. I think I actually prefer the original `sender` and `receiver`, they're slightly more cryptic but they're conventional which makes them instantly recognizable.

Reply:
I disagree on that, but I'll guess it's a taste/habit situation. Plus I don't like having a variable named after its type.

I'm open to changing it back if others feel the same way

Reply:
Your reason for making the change is good, I don't object to keeping it.